backend/cmm_helpers.ml

```diff
@@ -2059,18 +2067,50 @@ let ptr_offset ptr offset dbg =
 let direct_apply lbl args (pos, _mode) dbg =
   Cop (Capply (typ_val, pos), Cconst_symbol (lbl, dbg) :: args, dbg)
 
+let call_caml_apply mut clos args pos mode dbg =
+  let arity = List.length args in
+  let really_call_caml_apply clos args =
+    let cargs =
+      (Cconst_symbol (apply_function_sym arity mode, dbg) :: args) @ [clos]
+    in
+    Cop (Capply (typ_val, pos), cargs, dbg)
+  in
+  if !Flambda_backend_flags.caml_apply_inline_fast_path
+  then
+    (*    (if (= clos.arity N)
+     *      (app clos.direct a1 ... aN clos)
+     *      (app call_caml_applyN a1 .. aN clos)
+     *)
+    bind_list "arg" args (fun args ->
+        bind "fun" clos (fun clos ->
+            Cifthenelse
+              ( Cop
+                  ( Ccmpi Ceq,
+                    [ Cop
+                        ( Casr,
+                          [ get_field_gen mut clos 1 dbg;
+                            Cconst_int (pos_arity_in_closinfo, dbg) ],
+                          dbg );
+                      Cconst_int (arity, dbg) ],
+                    dbg ),
+                dbg,
+                Cop
+                  ( Capply (typ_val, pos),
+                    (get_field_gen mut clos 2 dbg :: args) @ [clos],
+                    dbg ),
+                dbg,
+                really_call_caml_apply clos args,
```

Comment:
Not sure this is something to address
now, leave to a `CR-someday`, or ignore,
but I would expect that call to be to a
function where the test is gone.

Reply:
We can do it if we enable this optimization unconditionally everywhere (and remove the new compiler flag and check). The current patch allows files compiled differently (with and without the new flag) to be linked together and `caml_applyN` is only generated at link time, so we still need it to have the fast path.

Reply:
I don't follow; my point was to have two versions
of `caml_applyN`: the existing one and the one
with only the cold path.

Reply:
ah, I missed the "two versions" part. I'll add a CR.